### PR TITLE
add end to end test for zipkin exporter

### LINKED ISSUE #1110: Create an end-to-end test for the zipkin exporter
Can use https://github.com/openzipkin/zipkin/tree/master/zipkin-junit for the testing.

### PR DESCRIPTION
**Pull request by jarebudev**
resolves #1110

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -97,6 +97,7 @@ subprojects {
         protobufVersion = '3.11.4'
         protocVersion = '3.11.4'
         zipkinReporterVersion = '2.12.2'
+        zipkinVersion = '2.18.3'
```
Comment (jarebudev) [inline build.gradle:100]: FYI, this isn't the latest version of ```zipkin-junit```. 
Any version of zipkin-junit newer than ```2.18.3``` use okhttp 4, which requires jvm >= 8

**Comment by carlosalberto** [comment]
Not super familiar with Zipkin, but overall looks great!

File: contrib/trace_propagators/build.gradle [hunk 2]
```diff
@@ -12,6 +12,8 @@ ext.moduleName = "io.opentelemetry.contrib.trace.propagation"
 dependencies {
     api project(':opentelemetry-api')
 
+    testImplementation libraries.jaeger_client
```
Comment (bogdandrutu) [inline contrib/trace_propagators/build.gradle:15]: I see what you did, please split this change in a separate PR. Also nice catch :)

File: exporters/zipkin/build.gradle [hunk 3]
```diff
@@ -19,11 +19,22 @@ dependencies {
     implementation libraries.zipkin_urlconnection
 
     testImplementation libraries.guava
-
+    testImplementation libraries.zipkin_junit
```
Comment (bogdandrutu) [inline exporters/zipkin/build.gradle:22]: nit: you can use comma separated here instead of having multiple `testImplementation`.

File: contrib/trace_propagators/build.gradle (hunk 2 above)
Comment (jarebudev) [reply contrib/trace_propagators/build.gradle:15]: Sure thing, I've created #1177.  As soon as that one is merged I'll rebase this PR.

File: build.gradle (hunk 1 above)
Comment (jkwatson) [reply build.gradle:103]: Are we losing test functionality by using the older version? Should we start thinking about having integration tests that use java 8 for the testing?

**Comment by jkwatson** [comment]
@adriancole not sure if you have any feedback on this, but it would be welcome!

File: build.gradle (hunk 1 above)
Comment (adriancole) [reply build.gradle:103]: yeah it seems odd to not use java 8 in tests. Most projects have build requirements at least 8 if not above these days.. consider when you want to add features that conditionally work with futures etc.

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    assertThat(resultCode).isEqualTo(SpanExporter.ResultCode.SUCCESS);
+    List<zipkin2.Span> zipkinSpans = zipkin.getTrace(TRACE_ID);
+
+    assertThat(zipkinSpans).isNotNull();
+    assertThat(zipkinSpans.size()).isEqualTo(1);
+    assertThat(zipkinSpans.get(0))
+        .isEqualTo(ZipkinSpanExporter.generateSpan(spanData, localEndpoint));
+  }
+
+  private static SpanData.Builder buildStandardSpan() {
```
Comment (adriancole) [inline exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:177]: typically when I write tests like this, I add a second method directly over and under it so that you can see the translation. This helps people understand the intent and any surprising things such as annotation naming choices or missed timestamp resolutions.

to do that, I would move attributes and annotations to locals here so that you can see everything without too much scrolling. the ID constants which are fine to be up top and plus you'd use these for zipkin also.

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    return SpanData.newBuilder()
+        .setTraceId(TraceId.fromLowerBase16(TRACE_ID, 0))
+        .setSpanId(SpanId.fromLowerBase16(SPAN_ID, 0))
+        .setParentSpanId(SpanId.fromLowerBase16(PARENT_SPAN_ID, 0))
+        .setTraceFlags(TraceFlags.builder().setIsSampled(true).build())
+        .setStatus(Status.OK)
+        .setKind(Span.Kind.SERVER)
+        .setHasRemoteParent(true)
+        .setName("Recv.helloworld.Greeter.SayHello")
+        .setStartEpochNanos(1505855794_194009601L)
```
Comment (adriancole) [inline exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:187]: interesting using nanos resolution, but will leave that aside..

when you add a method that generates the zipkin span, probably moving these timestamps to constants will help, and show that in zipkin they are /1000

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            .build();
+
+    exportAndVerify(configuration);
+  }
+
+  @Test
+  public void testExportFailedAsWrongEncoderUsed() {
+
+    ZipkinExporterConfiguration configuration =
+        ZipkinExporterConfiguration.builder()
```
Comment (DotSpy) [inline exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:137]:
```
        ZipkinExporterConfiguration.builder()
            .setSender(
                URLConnectionSender.newBuilder()
                    .endpoint(zipkin.httpUrl() + ENDPOINT_V2_SPANS)
                    .encoding(Encoding.JSON)
                    .build())
            .setServiceName(SERVICE_NAME)
            .setEncoder(SpanBytesEncoder.PROTO3)
            .build();
```
looking like it can be method with 4 parameters instead of duplicating it

File: build.gradle (hunk 1 above)
Comment (devinsba) [reply build.gradle:103]: I agree with Adrian here, I see no reason that **all** tests shouldn't build with at least Java 8.

Comment (jkwatson) [reply build.gradle:103]: I agree, but that's work for a different PR.

Comment (jarebudev) [reply build.gradle:103]: I've raised #1187 to cover this

**Comment by jkwatson** [comment]
looks like this needs a rebase.

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java (hunk 6 above)
Comment (jarebudev) [reply exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java]: While tidying the test up I extracted this into its own method

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java (hunk 4 above)
Comment (jarebudev) [reply exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:154]: I've not gone overboard on the tidying up, but I added a method to build the Zipkin span underneath the OTel span

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        .setStatus(Status.OK)
+        .setKind(Kind.SERVER)
+        .setHasRemoteParent(true)
+        .setName("Recv.helloworld.Greeter.SayHello")
+        .setStartEpochNanos(1505855794_194009601L)
+        .setAttributes(attributes)
+        .setTotalAttributeCount(attributes.size())
+        .setTimedEvents(annotations)
+        .setLinks(Collections.<SpanData.Link>emptyList())
+        .setEndEpochNanos(1505855799_465726528L)
```
Comment (adriancole) [inline exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:169]: for the purpose of "find", etc either make this a constant or use the same form as below? `1505855799000000L + 465726528L` using the same value representation helps see what's different in the formats

Comment (jarebudev) [reply exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java]: thanks, i've extracted these into constants now, should be good to go :)

File: exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    return Span.newBuilder()
+        .traceId(TRACE_ID)
+        .parentId(PARENT_SPAN_ID)
+        .id(SPAN_ID)
+        .kind(Span.Kind.SERVER)
+        .name(SPAN_NAME)
+        .timestamp(START_EPOCH_NANOS / 1000)
+        .duration((END_EPOCH_NANOS / 1000) - (START_EPOCH_NANOS / 1000))
+        .localEndpoint(localEndpoint)
+        .addAnnotation(RECEIVED_TIMESTAMP_NANOS / 1000, "RECEIVED")
```
Comment (adriancole) [inline exporters/zipkin/src/test/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporterEndToEndHttpTest.java:188]: looks good. actually the annotations themselves we can revisit, but in any case this E to E makes future discussions easier.
